### PR TITLE
feat: scenario-runner UI deploy flag + color-aware test suite (HTTP/DB/NR) + MSSQL CI-test fixes

### PR DESCRIPTION
**Pull request by galabastro**
🚀 Pull Request
========================

🎯 PR Type
----------
-   [x] **Feature:** Adds new functionality or user-facing change.
-   [x] **Bug Fix:** Solves a defect or incorrect behavior.
-   [ ] **Refactor:** Improves code structure/readability without changing external behavior.
-   [ ] **Documentation:** Updates to README, Wiki, or internal docs.
-   [x] **Chore:** Non-code changes (e.g., CI configuration, dependency bumps).

1\. Issue Tracking & Reference
------------------------------
-   **Ticket Number(s):** N/A
-   **Ticket Link(s):** N/A

2\. Summary of Changes
----------------------
Adds a deploy-time flag to gate the Scenario Runner web UI (while keeping its API reachable), makes the ReliBank test suite color-aware (blue/green) end-to-end, and fixes the MSSQL out-of-disk condition plus CI database reachability that the color-aware tests depend on.
-   **Expected Outcome:** With `scenario_ui_enabled=false`, `GET /scenario-runner/home` returns 404 while `/scenario-runner/api/*` still works. The test suite validates the specific deployed color across HTTP, DB, and New Relic (auto-resolving to the active color on the scheduled cron). MSSQL no longer fills its volume (SIMPLE recovery + 8Gi), and CI reaches the in-cluster DB privately without exposing it.

### Detailed Change Log & Justification
-   **Change 1:** Add `SCENARIO_UI_ENABLED` deploy flag — gate `read_root()` in `scenario_service/scenario_service.py` (404 when off), thread `scenario_ui_enabled` through `terraform/aks/app_module` + blue/green modules into the `infrastructure-config` ConfigMap, and expose it as a `Deploy ReliBank` workflow input.
    -   **Justification:** Hide the control-panel webpage per environment/color while keeping the scenario API reachable for backend services and automation.
-   **Change 2:** Color-directed HTTP + DB testing — new `tests/conftest.py` attaches `X-Test-Env: <color>` to every request; `test-suite.yml` resolves an effective color (`target_color` input, else active color read from `main-ingress`) and `kubectl port-forward`s that color's MSSQL to `127.0.0.1`.
    -   **Justification:** Tests must exercise and read back the specific deployed color, not a blue/green aggregate; the cron run needs no input and validates the active color.
-   **Change 3:** Color-scope New Relic queries — `tests/nrql_color.py` emits a namespace/`deploy.color` `WHERE` clause per event type (Span `k8s.namespace.name`, Log `namespace_name`, APM `deploy.color`); a shared one-liner in `utils/process_headers.py` stamps `deploy.color` from a `DEPLOY_COLOR` env injected on every service in `app_module/main.tf`.
    -   **Justification:** APM `Transaction` events carry no namespace attribute, so a custom `deploy.color` attribute (added once in shared code, zero per-service duplication) is required to color-scope APM NR checks.
-   **Change 4:** MSSQL reliability — `RECOVERY SIMPLE` in `transaction_service/mssql/init.sql` and PVC `2Gi → 8Gi` in `app_module/main.tf` + `k8s/base/databases/mssql-deployment.yaml`.
    -   **Justification:** Under FULL recovery with no log backups the transaction log grew unbounded and filled the 2Gi volume, causing `Error 1105/1101` and stalling DB-dependent flows/tests.
-   **Change 5:** CI DB reachability — `test-suite.yml` reads DB creds convention-agnostically (`DB_SERVER || 127.0.0.1`, reuse `MSSQL_SA_*`) and tunnels to the DB via port-forward instead of a public endpoint.
    -   **Justification:** Terraform environments keep MSSQL as a headless ClusterIP; a GitHub-hosted runner cannot reach it publicly, and exposing an SA-auth SQL Server publicly is undesirable.
-   **Change 6:** Scenario-UI webpage test — `tests/test_scenario_service.py::test_scenario_ui_flag_controls_webpage` asserts `/scenario-runner/home` matches `SCENARIO_UI_ENABLED` (200 + page marker vs 404) with the API always 200, color-directed via `X-Test-Env`.
    -   **Justification:** Automated verification that the flag actually controls the page per color, before cutover.
-   **Change 7:** Documentation — `docs/deployer/runbook.md` (UI toggle, MSSQL out-of-disk, DB-from-CI), `docs/deployer/deployer_primer.md` cheat sheet, `README.md`, `scenario_service/README.md`.
    -   **Justification:** Operators need the flag, the DB remediation steps, and the color-aware testing behavior documented.

3\. Testing
-----------
Testing was performed by running the **Relibank Test Suite** GitHub Actions workflow against the sandbox environment on this branch (the suite requires cluster access for the private DB port-forward and New Relic API keys, so it runs in CI rather than purely locally), supplemented by direct NerdGraph (NRQL) probes and public-URL curls.

Run the test suite locally with this in-development branch, and upload the results. If new tests are required, like for new features, add those tests and call them out here.
```cd tests && ./run_tests.sh > ../../test_results.txt```

New tests added: `tests/test_scenario_service.py::test_scenario_ui_flag_controls_webpage`, shared `tests/conftest.py` (color-directed session), `tests/nrql_color.py` (NR color-scoping helper).

### Testing Strategy
-   **Environment Used:** Sandbox (`relibank-sandbox`, active color blue).
-   **Production Safety Assessment:** `scenario_ui_enabled` defaults to `true` (no behavior change unless explicitly set); `DEPLOY_COLOR`/`deploy.color` is additive telemetry; the MSSQL and CI changes are validated on sandbox and gated by env vars/inputs. The Terraform changes apply to Terraform-deployed environments (sandbox/staging/analysts); legacy prod/`events` uses the separate `k8s/base` path and its existing `DB_SERVER` secret (unchanged, handled via `||` fallback).
-   **What areas were NOT tested and why?** Browser (`PageView`) and mssql-collector `Metric` color-scoping — those event types carry no color dimension and remain intentionally env-level. The prod/`events` deploy path was not exercised (out of scope; different deployment mechanism).

### Coverage Details

| Scenario      | Details                                                                    | Results                           |
| ------------- | -------------------------------------------------------------------------- | --------------------------------- |
| Happy Path    | Deploy blue with `scenario_ui_enabled=false`; run full suite color-directed to blue. `/scenario-runner/home` → 404, `/scenario-runner/api/scenarios` → 200. | Pass. 15/17 modules green; `test_scenario_service` (incl. webpage-flag test) green; `test_newrelic_instrumentation` and `test_db_pool_e2e` pass and are color-scoped. |
| Edge Case 1   | Cron / no `target_color` input — effective color must auto-resolve to the active color from `main-ingress`. | Pass. Resolved `Effective color: blue`; tunneled `relibank-blue`; tests scoped to active color. |
| Edge Case 2   | Color isolation — same query for the non-active color must return no data. | Pass. `deploy.color='blue'` → 88 txns (pool-a, pool-b); `deploy.color='green'` → 0. |
| Edge Case 3   | APM `Transaction` has no namespace attribute (regressed 2 tests); fixed via `deploy.color` custom attribute. | Pass. `deploy.color` appeared on APM (218 txns blue); `test_newrelic_instrumentation` + `test_db_pool_e2e` green after fix. |
| Edge Case 4   | Pre-existing New Relic ingestion-lag / HTTP read-timeout / Metric pytest-timeout tests (out of scope for this PR). | Fail (unchanged). `test_newrelic_risk_assessment`, `test_nrdot_mssql_collector` — NR-lag/NoneType/Metric-timeout, not color-related. |

4\. Evidence
------------
-   **Screenshots/Gifs:** 
image

-   **Console/Log Output Snippets:**
    - Scenario UI flag (public URL, blue, flag off): `GET /scenario-runner/home` → `404`; `GET /scenario-runner/api/scenarios` → `200`; `GET /scenario-runner` → `200`; `GET /` → `200`.
    - Effective-color resolution (cron path, no input): `Effective color: blue`; `MSSQL tunnel up (relibank-blue)`.
    - APM color attribute: `SELECT count(*) FROM Transaction WHERE deploy.color IS NOT NULL FACET deploy.color` → `blue=218`.
    - Color isolation: `... WHERE appName LIKE '%Accounts Service%' AND db.pool_id IS NOT NULL AND deploy.color='blue'` → `count 88, uniques(db.pool_id)=[pool-a, pool-b]`; same query for `deploy.color='green'` → `count 0`.
-   **Test Report Links:** Relibank Test Suite (sandbox, this branch): https://github.com/newrelic/relibank/actions/runs/28979937323 ; scenario-only run (webpage flag, 16 passed): https://github.com/newrelic/relibank/actions/runs/28906597189

_Testing and evidence are **REQUIRED** before requesting review!_

5\. Review and Merge Checklist
------------------------------
Confirm the following prerequisites have been met by checking the boxes:
-   [x] All blocking review comments from the latest round are **resolved**.
-   [x] This branch has been rebased against the `main` branch and all **merge conflicts are resolved**.
-   [x] The code includes sufficient comments, particularly in complex or non-obvious areas.
-   [x] The code adheres to project **coding standards** (linting, style guides, best practices).
-   [x] Relevant documentation and runbooks have been updated, if necessary.
-   [x] My changes generate no new warnings or errors.
-   [x] My commits are squashed into a single, descriptive commit.
-   [x] My test results are uploaded as a text file, and new tests were created where required.

### CODE REVIEW

**Comment by galabastro** [comment]
```
Relibank Test Suite — sandbox (branch: ftr/scenario-ui-flag)
GitHub Actions run: https://github.com/newrelic/relibank/actions/runs/28979937323
target_environment=sandbox   test_suite=all   effective color=blue (auto-resolved)
================================================================================

SUMMARY
  Python modules: 15 passed, 2 failed (out-of-scope New Relic latency/ingestion bucket)
  Frontend: 37 passed (5 files)

PYTHON TESTS (per module)
--------------------------------------------------------------------------------
Running test_ab_testing_scenarios.py...
============================= 10 passed in 19.77s ==============================
✓ test_ab_testing_scenarios.py passed
Running test_apm_user_tracking.py...
============================= 15 passed in 22.89s ==============================
✓ test_apm_user_tracking.py passed
Running test_browser_user_tracking.py...
============================== 7 passed in 4.17s ===============================
✓ test_browser_user_tracking.py passed
Running test_card_payment_transactions.py...
========================= 4 passed in 77.13s (0:01:17) =========================
✓ test_card_payment_transactions.py passed
Running test_db_pool_e2e.py...
======================== 1 passed in 144.92s (0:02:24) =========================
✓ test_db_pool_e2e.py passed
Running test_db_pool_scenarios.py...
============================== 7 passed in 8.22s ===============================
✓ test_db_pool_scenarios.py passed
Running test_end_to_end.py...
============================= 23 passed in 39.50s ==============================
✓ test_end_to_end.py passed
Running test_mfe_single.py...
======================== 1 passed in 225.58s (0:03:45) =========================
✓ test_mfe_single.py passed
Running test_microfrontend_telemetry.py...
======================== 4 passed in 504.45s (0:08:24) =========================
✓ test_microfrontend_telemetry.py passed
Running test_newrelic_instrumentation.py...
=================== 5 passed, 1 skipped in 184.88s (0:03:04) ===================
✓ test_newrelic_instrumentation.py passed
Running test_newrelic_risk_assessment.py...
======================== 4 failed in 193.77s (0:03:13) =========================
✗ test_newrelic_risk_assessment.py failed
Running test_nrdot_mssql_collector.py...
======================== 5 errors in 300.10s (0:05:00) =========================
✗ test_nrdot_mssql_collector.py failed
Running test_payment_scenarios.py...
======================== 5 passed in 129.26s (0:02:09) =========================
✓ test_payment_scenarios.py passed
Running test_rogue_deployment_scenarios.py...
============================== 7 passed in 49.86s ==============================
✓ test_rogue_deployment_scenarios.py passed
Running test_scenario_service.py...
============================== 16 passed in 5.73s ==============================
✓ test_scenario_service.py passed
Running test_stress_scenarios.py...
============================== 1 passed in 4.56s ===============================
✓ test_stress_scenarios.py passed
Running test_with_interactions.py...
======================== 2 passed in 252.07s (0:04:12) =========================
✓ test_with_interactions.py passed

FAILURE DETAIL — the 2 failing modules are the known, out-of-scope bucket
--------------------------------------------------------------------------------
FAILED test_newrelic_risk_assessment.py::test_risk_assessment_logs_in_newrelic - TypeError: 'NoneType' object is not subscriptable
FAILED test_newrelic_risk_assessment.py::test_declined_payment_shows_in_newrelic - AssertionError: Expected declined payment logs, found 0
FAILED test_newrelic_risk_assessment.py::test_agent_model_tracking_in_newrelic - AssertionError: Expected agent model tracking in logs, found 0
FAILED test_newrelic_risk_assessment.py::test_risk_assessment_ebpf_traces - TypeError: unsupported format string passed to NoneType.__format__
ERROR test_nrdot_mssql_collector.py::test_lock_contention_metrics - Failed: Timeout (>300.0s) from pytest-timeout.
ERROR test_nrdot_mssql_collector.py::test_connection_and_throughput_metrics - Failed: Timeout (>300.0s) from pytest-timeout.
ERROR test_nrdot_mssql_collector.py::test_buffer_pool_metrics - Failed: Timeout (>300.0s) from pytest-timeout.
ERROR test_nrdot_mssql_collector.py::test_sql_compilation_metrics - Failed: Timeout (>300.0s) from pytest-timeout.
ERROR test_nrdot_mssql_collector.py::test_wait_stats_metrics - Failed: Timeout (>300.0s) from pytest-timeout.

FRONTEND TESTS
--------------------------------------------------------------------------------
Test Files  5 passed (5)
      Tests  37 passed (37)
```

Definitive confirmation — these failures are pre-existing on main, not introduced by this branch. Both modules fail identically on:

main @ 2026-07-02 (run 28594218643) — before this branch existed (branch's first commit was ~July 6)
main @ 2026-07-09 (run 29023337246) — current main

**Comment by galabastro** [comment]
🚀 Pull Request (Updated)
========================

🎯 PR Type
----------
-   [x] **Feature:** Adds new functionality or user-facing change.
-   [x] **Bug Fix:** Solves a defect or incorrect behavior.
-   [ ] **Refactor:** Improves code structure/readability without changing external behavior.
-   [ ] **Documentation:** Updates to README, Wiki, or internal docs.
-   [x] **Chore:** Non-code changes (e.g., CI configuration, dependency bumps).

1\. Issue Tracking & Reference
------------------------------
-   **Ticket Number(s):** N/A
-   **Ticket Link(s):** N/A

2\. Summary of Changes
----------------------
Gates the Scenario Runner web UI by deploy environment (hidden in `prod`, shown elsewhere) while keeping its API reachable, makes the ReliBank test suite color-aware (blue/green) end-to-end, and fixes the MSSQL out-of-disk condition plus CI database reachability that the color-aware tests depend on.
-   **Expected Outcome:** In `prod`, `GET /scenario-runner/home` returns 404 (UI hidden); in every other environment it serves the page — derived automatically from the environment, no workflow toggle. `/scenario-runner/api/*` works regardless. The test suite validates the specific deployed color across HTTP, DB, and New Relic (auto-resolving to the active color on the scheduled cron). MSSQL no longer fills its volume (SIMPLE recovery + 8Gi), and CI reaches the in-cluster DB privately without exposing it.

### Detailed Change Log & Justification
-   **Change 1:** Gate the Scenario Runner web UI by environment — `read_root()` in `scenario_service/scenario_service.py` returns 404 when `SCENARIO_UI_ENABLED` is off; the value is **derived in Terraform from the deploy environment** (`SCENARIO_UI_ENABLED = tostring(var.demo_environment != "prod")`) into the `infrastructure-config` ConfigMap — hidden in `prod`, shown everywhere else. No workflow input.
    -   **Justification:** Hide the control-panel webpage in prod while keeping it available elsewhere and the scenario API always reachable — automatic, with one fewer workflow-dropdown option to get wrong.
-   **Change 2:** Color-directed HTTP + DB testing — new `tests/conftest.py` attaches `X-Test-Env: <color>` to every request; `test-suite.yml` resolves an effective color (`target_color` input, else active color read from `main-ingress`) and `kubectl port-forward`s that color's MSSQL to `127.0.0.1`.
    -   **Justification:** Tests must exercise and read back the specific deployed color, not a blue/green aggregate; the cron run needs no input and validates the active color.
-   **Change 3:** Color-scope New Relic queries — `tests/nrql_color.py` emits a namespace/`deploy.color` `WHERE` clause per event type (Span `k8s.namespace.name`, Log `namespace_name`, APM `deploy.color`); a shared one-liner in `utils/process_headers.py` stamps `deploy.color` from a `DEPLOY_COLOR` env injected on every service in `app_module/main.tf`.
    -   **Justification:** APM `Transaction` events carry no namespace attribute, so a custom `deploy.color` attribute (added once in shared code, zero per-service duplication) is required to color-scope APM NR checks.
-   **Change 4:** MSSQL reliability — `RECOVERY SIMPLE` in `transaction_service/mssql/init.sql` and PVC `2Gi → 8Gi` in `app_module/main.tf` + `k8s/base/databases/mssql-deployment.yaml`.
    -   **Justification:** Under FULL recovery with no log backups the transaction log grew unbounded and filled the 2Gi volume, causing `Error 1105/1101` and stalling DB-dependent flows/tests.
-   **Change 5:** CI DB reachability — `test-suite.yml` reads DB creds convention-agnostically (`DB_SERVER || 127.0.0.1`, reuse `MSSQL_SA_*`) and tunnels to the DB via port-forward instead of a public endpoint.
    -   **Justification:** Terraform environments keep MSSQL as a headless ClusterIP; a GitHub-hosted runner cannot reach it publicly, and exposing an SA-auth SQL Server publicly is undesirable.
-   **Change 6:** Scenario-UI webpage test — `tests/test_scenario_service.py::test_scenario_ui_flag_controls_webpage` asserts `/scenario-runner/home` matches `SCENARIO_UI_ENABLED` (200 + page marker vs 404) with the API always 200, color-directed via `X-Test-Env`.
    -   **Justification:** Automated verification that the flag actually controls the page per color, before cutover.
-   **Change 7:** Documentation — `docs/deployer/runbook.md` (UI toggle, MSSQL out-of-disk, DB-from-CI), `docs/deployer/deployer_primer.md` cheat sheet, `README.md`, `scenario_service/README.md`.
    -   **Justification:** Operators need the flag, the DB remediation steps, and the color-aware testing behavior documented.

3\. Testing
-----------
Testing was performed by running the **Relibank Test Suite** GitHub Actions workflow against the sandbox environment on this branch (the suite requires cluster access for the private DB port-forward and New Relic API keys, so it runs in CI rather than purely locally), supplemented by direct NerdGraph (NRQL) probes and public-URL curls.

Run the test suite locally with this in-development branch, and upload the results. If new tests are required, like for new features, add those tests and call them out here.
```cd tests && ./run_tests.sh > ../../test_results.txt```

New tests added: `tests/test_scenario_service.py::test_scenario_ui_flag_controls_webpage`, shared `tests/conftest.py` (color-directed session), `tests/nrql_color.py` (NR color-scoping helper).

### Testing Strategy
-   **Environment Used:** Sandbox (`relibank-sandbox`, active color blue).
-   **Production Safety Assessment:** the scenario-UI gate is env-derived (`prod → hidden`, else shown) — no behavior change to non-prod runtime beyond the intended UI visibility; `DEPLOY_COLOR`/`deploy.color` is additive telemetry; the MSSQL and CI changes are validated on sandbox. The Terraform changes apply to Terraform-deployed environments (sandbox/staging/analysts); legacy prod/`events` uses the separate `k8s/base` path and its existing `DB_SERVER` secret (unchanged, handled via `||` fallback).
-   **What areas were NOT tested and why?** Browser (`PageView`) and mssql-collector `Metric` color-scoping — those event types carry no color dimension and remain intentionally env-level. The prod/`events` deploy path was not exercised (out of scope; different deployment mechanism).

### Coverage Details

| Scenario      | Details                                                                    | Results                           |
| ------------- | -------------------------------------------------------------------------- | --------------------------------- |
| Happy Path    | Deploy sandbox (non-prod → UI shown); run full suite color-directed to the deployed color. `/scenario-runner/api/scenarios` → 200; `/scenario-runner/home` state matches the env rule (shown in non-prod). | Pass. 15/17 modules green; `test_scenario_service` (incl. env-derived webpage-flag test) green; `test_newrelic_instrumentation` and `test_db_pool_e2e` pass and are color-scoped. |
| Edge Case 1   | Cron / no `target_color` input — effective color must auto-resolve to the active color from `main-ingress`. | Pass. Resolved `Effective color: blue`; tunneled `relibank-blue`; tests scoped to active color. |
| Edge Case 2   | Color isolation — same query for the non-active color must return no data. | Pass. `deploy.color='blue'` → 88 txns (pool-a, pool-b); `deploy.color='green'` → 0. |
| Edge Case 3   | APM `Transaction` has no namespace attribute (regressed 2 tests); fixed via `deploy.color` custom attribute. | Pass. `deploy.color` appeared on APM (218 txns blue); `test_newrelic_instrumentation` + `test_db_pool_e2e` green after fix. |
| Edge Case 4   | Pre-existing New Relic ingestion-lag / HTTP read-timeout / Metric pytest-timeout tests (out of scope for this PR). | Fail (unchanged). `test_newrelic_risk_assessment`, `test_nrdot_mssql_collector` — NR-lag/NoneType/Metric-timeout, not color-related. |

4\. Evidence
------------
-   **Screenshots/Gifs:** N/A
-   **Console/Log Output Snippets:**
    - Scenario UI flag (public URL, blue, flag off): `GET /scenario-runner/home` → `404`; `GET /scenario-runner/api/scenarios` → `200`; `GET /scenario-runner` → `200`; `GET /` → `200`.
    - Effective-color resolution (cron path, no input): `Effective color: blue`; `MSSQL tunnel up (relibank-blue)`.
    - APM color attribute: `SELECT count(*) FROM Transaction WHERE deploy.color IS NOT NULL FACET deploy.color` → `blue=218`.
    - Color isolation: `... WHERE appName LIKE '%Accounts Service%' AND db.pool_id IS NOT NULL AND deploy.color='blue'` → `count 88, uniques(db.pool_id)=[pool-a, pool-b]`; same query for `deploy.color='green'` → `count 0`.
-   **Test Report Links:** Relibank Test Suite (sandbox, this branch): https://github.com/newrelic/relibank/actions/runs/28979937323 ; scenario-only run (webpage flag, 16 passed): https://github.com/newrelic/relibank/actions/runs/28906597189

_Testing and evidence are **REQUIRED** before requesting review!_

5\. Review and Merge Checklist
------------------------------
Confirm the following prerequisites have been met by checking the boxes:
-   [x] All blocking review comments from the latest round are **resolved**.
-   [x] This branch has been rebased against the `main` branch and all **merge conflicts are resolved**.
-   [x] The code includes sufficient comments, particularly in complex or non-obvious areas.
-   [x] The code adheres to project **coding standards** (linting, style guides, best practices).
-   [x] Relevant documentation and runbooks have been updated, if necessary.
-   [x] My changes generate no new warnings or errors.
-   [x] My commits are squashed into a single, descriptive commit.
-   [x] My test results are uploaded as a text file, and new tests were created where required.

File: .github/workflows/deploy-relibank.yml [hunk 1]
```diff
@@ -275,3 +275,6 @@ jobs:
     with:
       target_environment: ${{ github.event.inputs.environment }}
       test_suite: all
+      # Validate the just-deployed color before cutover (via X-Test-Env). The scenario-UI
+      # expected state is derived from the environment inside test-suite.yml (prod -> off).
```
Comment (jgoddard19) [inline .github/workflows/deploy-relibank.yml:279]: This scenario-UI part is probably not needed

**Comment by jgoddard19** [comment]
Looks good - make sure the deploy and testing runbooks/docs are updated with the new test-suite color changes

**Review by jgoddard19 — APPROVED** [review]
